### PR TITLE
[Optimization] xgrammar async compile, multi thread, speed up

### PR DESCRIPTION
**Pull request by ST-XX**
## Motivation
Optimize the integration of xgrammar by introducing asynchronous compilation and native caching. Improve efficiency for CUDA platforms with inplace operations and DLPack interconversion, and remove redundant backend caching logic.
优化 xgrammar 的集成，采用异步编译与原生缓存机制，提升 CUDA 平台效率，同时去除冗余的后端缓存逻辑。

## Modifications

- Refactored xgrammar to use asynchronous `compile` and implemented native caching.  
- Removed caching from the backend to avoid duplication.  
- Triggered xgrammar compilation during the Prefill stage, and joined the compile result before sampling the first token in decode.  
- For CUDA platforms:  
  - Used DLPack as an intermediate format for conversion between `paddle.Tensor` and `torch.Tensor`.  
  - Leveraged CUDA hardware for inplace acceleration of `xgr.apply_token_bitmask_inplace`.  
  - Removed previous GPU to CPU numpy conversion.  
- Other platforms retain existing logic.

- xgrammar 改为异步 `compile` 并实现了原生缓存机制。  
- 去掉了 backend 中的缓存，避免重复。  
- Prefill 阶段发起 xgrammar 编译，在 decode 第一个 token 的 sampler 之前 join 编译结果。  
- 对于 CUDA 平台：  
  - 使用 DLPack 作为 paddle.Tensor 和 torch.Tensor 之间的中间转换格式。  
  - 利用 CUDA 硬件进行 inplace 加速 `xgr.apply_token_bitmask_inplace`。  
  - 移除了之前 GPU 到 CPU 的 numpy 转换。  
- 其他平台维持原有逻辑。
## Usage or Command

```python
import openai

port = "8170"
client = openai.Client(base_url=f"http://127.0.0.1:{port}/v1", api_key="null")

completion = client.chat.completions.create(
    model="null",
    messages=[
        {
            "role": "user",
            "content": "Generate a JSON object containing: names of China's Four Great Inventions, their dynasties of origin, and brief descriptions (each under 50 characters)",
        }
    ],
    response_format={"type": "json_object"}
)
print(completion.choices[0].message.content)
```

## Accuracy Tests

预期会影响结果，以符合受限解码的结构化要求。

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [x] Provide accuracy results.
- [x] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/model_executor/guided_decoding/xgrammar_backend.py [hunk 1]
```diff
@@ -155,23 +124,21 @@ def reset(self) -> None:
     def accept_token(self, token: int) -> None:
         """
         Validate and accept a generated token against the grammar constraints.
+        when accept eos_token, is_terminated = True
```
Comment (kevincheng2) [inline fastdeploy/model_executor/guided_decoding/xgrammar_backend.py:127]: 这里在哪里判断的eos_token啊？输出超长的场景怎么处理的？

File: fastdeploy/model_executor/guided_decoding/xgrammar_backend.py [hunk 2]
```diff
@@ -467,3 +440,49 @@ def schema_format(self, request: Request):
...
+        )
+        dlpack2 = torch.utils.dlpack.to_dlpack(t_logits)
+        return paddle.utils.dlpack.from_dlpack(dlpack2)
+    else:
+        origin_place = logits.place
+        origin_dtype = logits.dtype
+        logits = torch.from_numpy(logits.numpy())
+
+        logits = logits.float()  # cpu
+        apply_token_bitmask_inplace(
```
Comment (kevincheng2) [inline fastdeploy/model_executor/guided_decoding/xgrammar_backend.py:483]: 这个算子在多硬件上好像没有验证过？不确定能不能用

File: fastdeploy/model_executor/guided_decoding/xgrammar_backend.py [hunk 3]
```diff
@@ -467,3 +440,49 @@ def schema_format(self, request: Request):
...
+        indices (Optional[List[int]]): Optional list of batch indices to apply mask to
+
+    Returns:
+        paddle.Tensor: The modified logits tensor
+    """
+
+    if current_platform.is_cuda():
+        dlpack = paddle.utils.dlpack.to_dlpack(logits)
+        t_logits = torch.from_dlpack(dlpack)
+        apply_token_bitmask_inplace(
```
Comment (kevincheng2) [inline fastdeploy/model_executor/guided_decoding/xgrammar_backend.py:470]: 这个算子是支持paddle.tensor 的吧，为什么还要转torch.tensor 呢

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 4]
```diff
@@ -66,140 +67,167 @@ class GuidedDecoding:
...
-                for token in prefill_tokens:
-                    self.logits_processor[ids].accept_token(token)
-            else:
-                self.logits_processor[ids] = [future, prefill_tokens]
-
-    def update_vocab_mask(self, skip_idx_list: List[int] = []):
-        """update vocab mask. (cpu-heavy operation)"""
-        if len(self.logits_processor) == 0:
+        """add logits processor to SamplerProcessor"""
+        assert len(prefill_tokens) == 0
```
Comment (kevincheng2) [inline fastdeploy/model_executor/layers/sample/sampler.py]: PD分离场景下，prefill_tokens非空？

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 5]
```diff
@@ -66,140 +67,167 @@ class GuidedDecoding:
...
+                continue
+
+            if self.token_bitmask is None:
+                self.token_bitmask = self.logits_processors[idx].allocate_token_bitmask()
+
+            if self.should_fill_bitmask(idx):
                 processor.fill_token_bitmask(self.token_bitmask, idx)
 
-    def apply_token_mask(self, logits: paddle.Tensor, skip_idx_list: List[int] = []):
+    def apply_token_mask(self, logits: paddle.Tensor, prefill_done_idxs: List[int] = []):
```
Comment (kevincheng2) [inline fastdeploy/model_executor/layers/sample/sampler.py:169]: decode step间的异步是不是还没有加？

File: fastdeploy/worker/gcu_model_runner.py
Comment (kevincheng2) [inline fastdeploy/worker/gcu_model_runner.py:1]: 多硬件的场景还没验证过。如果要支持，优先支持xpu吧

File: fastdeploy/model_executor/layers/sample/sampler.py (hunk 4 above)
Comment (ST-XX) [reply fastdeploy/model_executor/layers/sample/sampler.py]: PD 分离场景还没验证。这里会 assert 挂掉

File: fastdeploy/worker/gcu_model_runner.py
Comment (ST-XX) [reply fastdeploy/worker/gcu_model_runner.py:1]: 这里因为接口变了，必须得同步修改。xpu ci 过了。

File: fastdeploy/model_executor/guided_decoding/xgrammar_backend.py (hunk 1 above)
Comment (ST-XX) [reply fastdeploy/model_executor/guided_decoding/xgrammar_backend.py:127]: eos accept 之后，matcher 的状态就是is_terminated，下面就会被重置掉了。后面输出的 token 不会再限制格式。开 ignore_eos 之后也可以继续生成。

File: fastdeploy/model_executor/guided_decoding/xgrammar_backend.py (hunk 2 above)
Comment (ST-XX) [reply fastdeploy/model_executor/guided_decoding/xgrammar_backend.py:483]: 这里是纯 cpu 操作。             bitmask=token_bitmask.to(logits.device, non_blocking=True),
  这个逻辑有点误导，实际 to 的还是 cpu

File: fastdeploy/model_executor/guided_decoding/xgrammar_backend.py (hunk 3 above)
Comment (ST-XX) [reply fastdeploy/model_executor/guided_decoding/xgrammar_backend.py:470]: 这里还是原生的 xgr. apply_token_bitmask_inplace 接口，只支持 tensor.Tensor

**Review by yuanlehome — DISMISSED** [review]

**Review by kevincheng2 — DISMISSED** [review]
